crates/re_web_viewer_server/src/lib.rs

```diff
@@ -191,73 +89,186 @@ impl FromStr for WebViewerServerPort {
+        }
     }
 }
 
-pub(crate) fn server_url(local_addr: &SocketAddr) -> String {
-    if local_addr.ip().is_unspecified() {
-        // "0.0.0.0"
-        format!("http://localhost:{}", local_addr.port())
-    } else {
-        format!("http://{local_addr}")
+impl Drop for WebViewerServer {
+    fn drop(&mut self) {
+        if let Some(thread_handle) = self.thread_handle.take() {
+            self.inner
+                .shutdown
+                .store(true, std::sync::atomic::Ordering::Release);
+            self.inner.server.unblock();
+            thread_handle.join().ok();
+        }
+    }
+}
+
+impl WebViewerServerInner {
+    fn serve(&self) {
+        loop {
+            let request = self.server.recv();
+            if self.shutdown.load(std::sync::atomic::Ordering::Acquire) {
+                return;
+            }
+
+            let request = match request {
+                Ok(request) => request,
+                Err(err) => {
+                    re_log::error!("Failed to receive http request: {err}");
+                    continue;
+                }
+            };
+
+            if let Err(err) = self.send_response(request) {
+                re_log::error!("Failed to send http response: {err}");
+            }
+        }
+    }
+
+    #[cfg(feature = "analytics")]
+    fn on_serve_wasm(&self) {
+        if let Some(analytics) = &self.analytics {
+            analytics.record(re_analytics::event::ServeWasm);
+        }
+    }
+
+    #[cfg(feature = "__ci")]
+    #[allow(clippy::needless_pass_by_value)]
+    fn send_response(&self, _request: tiny_http::Request) -> Result<(), std::io::Error> {
+        if false {
+            self.on_serve_wasm(); // to silence warning about the function being unused
+        }
+
+        // panic! is not enough in hyper (since it uses catch_unwind) - that only kills this thread. We want to quit.
+        eprintln!("web_server compiled with '__ci' feature (or `--all-features`). DON'T DO THAT! It's only for the CI!");
+        std::process::abort();
+    }
+
+    #[cfg(not(feature = "__ci"))]
+    fn send_response(&self, request: tiny_http::Request) -> Result<(), std::io::Error> {
+        // Strip arguments from url so we get the actual path.
+        let url = request.url();
+        let path = url.split('?').next().unwrap_or(url);
+
+        let (mime, bytes) = match path {
+            "/" | "/index.html" => ("text/html", data::INDEX_HTML),
+            "/favicon.svg" => ("image/svg+xml", data::FAVICON),
+            "/favicon.ico" => {
+                // We don't have a favicon.ico.
+                return request.respond(tiny_http::Response::empty(404));
+            }
+            "/sw.js" => ("text/javascript", data::SW_JS),
+            "/re_viewer.js" => ("text/javascript", data::VIEWER_JS),
+            "/re_viewer_bg.wasm" => {
+                #[cfg(feature = "analytics")]
+                self.on_serve_wasm();
+                ("application/wasm", data::VIEWER_WASM)
```

Comment:
We should add a TODO here about compression. We could easily generate a gziped .wasm in `re_build_web_server` and serve it here if the incoming request has `gzip` in a `Accept-Encoding` header